readthedocs/oauth/services/githubapp.py

```diff
@@ -0,0 +1,473 @@
+
+    @cached_property
+    def app_installation(self) -> GHInstallation:
+        """
+        Return the installation object from the GitHub API.
+
+        Usefull to interact with installation related endpoints.
+
+        If the installation is no longer accessible, this will raise a GithubException.
+        """
+        return self.gha_client.get_app_installation(
+            self.installation.installation_id,
+        )
+
+    @cached_property
+    def installation_client(self) -> Github:
+        """Return a client authenticated as the GitHub installation to interact with the GH API."""
+        return self.gha_client.get_github_for_installation(self.installation.installation_id)
+
+    @classmethod
+    def for_project(cls, project):
+        """
+        Return a GitHubAppService for the installation linked to the project.
+
+        Since this service only works for projects that have a remote repository,
+        and are linked to a GitHub App installation,
+        this returns only one service or None.
+        """
+        if not project.remote_repository or not project.remote_repository.github_app_installation:
+            return None
+
+        yield cls(project.remote_repository.github_app_installation)
+
+    @classmethod
+    def for_user(cls, user):
+        """
+        Return a GitHubAppService for each installation accessible to the user.
+
+        In order to get the installations accessible to the user, we need to use
+        the GitHub API authenticated as the user, making use of the user's access token
+        (not the installation token).
+
+        See:
+
+        - https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/authenticating-with-a-github-app-on-behalf-of-a-user
+        - https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/generating-a-user-access-token-for-a-github-app
+        - https://docs.github.com/en/rest/apps/installations?apiVersion=2022-11-28#list-app-installations-accessible-to-the-user-access-token
+
+        .. note::
+
+           If the installation wasn't in our database, we create it
+           (but we don't sync the repositories, since the caller should be responsible for that).
+
+        .. note::
+
+           User access tokens expire after 8 hours, but our OAuth2 client should handle refreshing the token.
+           But, the refresh token expires after 6 months, in order to refresh that token,
+           the user needs to sign in using GitHub again (just a normal sing-in, not a re-authorization or sign-up).
+        """
+        social_accounts = SocialAccount.objects.filter(
+            user=user,
+            provider=cls.allauth_provider.id,
+        )
+        for account in social_accounts:
+            oauth2_client = get_oauth2_client(account)
+            resp = oauth2_client.get("https://api.github.com/user/installations")
+            if resp.status_code != 200:
+                log.info(
```

Comment:
Is `info` the right level here? Seems like more of a warning? What are the cases we expect to happen here? Should we retry?

Reply:
This happens only if the user has revoked access to our app, maybe if GH is down, but this doesn't have side effects like deleting stuff.